crates/pathfinder/src/sync/p2p/state_updates.rs

```diff
@@ -0,0 +1,291 @@
+use std::collections::HashMap;
+
+use anyhow::Context;
+use p2p::PeerData;
+use pathfinder_common::{
+    state_update::ContractUpdates, BlockHash, BlockHeader, BlockNumber, StateUpdate,
+    StorageCommitment,
+};
+use pathfinder_crypto::Felt;
+use pathfinder_merkle_tree::{
+    contract_state::{update_contract_state, ContractStateUpdateResult},
+    StorageCommitmentTree,
+};
+use pathfinder_storage::{Node, Storage};
+use tokio::task::spawn_blocking;
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum ContractDiffSyncError {
+    #[error(transparent)]
+    DatabaseOrComputeError(#[from] anyhow::Error),
+    #[error("Storage commitment mismatch")]
+    StorageCommitmentMismatch(PeerData<BlockNumber>),
+}
+
+/// Returns the first block number whose state update is missing in storage, counting from genesis
+pub(super) async fn next_missing(
+    storage: Storage,
+    head: BlockNumber,
+) -> anyhow::Result<Option<BlockNumber>> {
+    spawn_blocking(move || {
+        let mut db = storage
+            .connection()
+            .context("Creating database connection")?;
+        let db = db.transaction().context("Creating database transaction")?;
+
+        if let Some(highest) = db
+            .highest_state_update()
+            .context("Querying highest state update")?
+        {
+            Ok((highest < head).then_some(highest + 1))
+        } else {
+            Ok(Some(BlockNumber::GENESIS))
+        }
+    })
+    .await
+    .context("Joining blocking task")?
+}
+
+#[derive(Debug)]
+pub(super) struct VerificationOk {
+    block_number: BlockNumber,
+    block_hash: BlockHash,
+    storage_commitment: StorageCommitment,
+    contract_update_results: Vec<ContractStateUpdateResult>,
+    trie_nodes: HashMap<Felt, Node>,
+    contract_updates: ContractUpdates,
+}
+
+pub(super) async fn verify(
+    storage: Storage,
+    contract_updates: Vec<PeerData<(BlockNumber, ContractUpdates)>>,
+    verify_trie_hashes: bool,
+) -> Result<Vec<PeerData<VerificationOk>>, ContractDiffSyncError> {
+    tokio::task::spawn_blocking(move || {
+        use rayon::prelude::*;
+
+        contract_updates
+            .into_par_iter()
+            .map(|x| verify_one(storage.clone(), x, verify_trie_hashes))
+            .collect::<Result<Vec<_>, _>>()
```

Comment:
You cannot do this because this includes state trie updates - these have to be sequentially applied

Reply:
Good catch!